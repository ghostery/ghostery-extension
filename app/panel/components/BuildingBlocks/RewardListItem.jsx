/**
 * Reward List Item Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import { Link } from 'react-router-dom';
import ClassNames from 'classnames';
import { computeTimeDelta } from '../../utils/utils';

/**
 * @class Implements a single reward in a list for the Rewards Panel
 * @memberof PanelBuildingBlocks
 */
class RewardListItem extends React.Component {
	constructor(props) {
		super(props);
		this.state = {};

		// Event Bindings
		this.handleClick = this.handleClick.bind(this);
		this.clickCloseButton = this.clickCloseButton.bind(this);

		// Send Signal
		this.props.actions.sendSignal('offer_shown', this.props.id);
	}

	/**
	 * Handle the click event: everything normal unless it is disabled
	 * @param  {Object} event the click event
	 */
	handleClick(event) {
		const { disabled, unread, id } = this.props;
		if (disabled) {
			event.preventDefault();
		} else {
			if (unread) {
				this.props.actions.sendSignal('offer_click_specific_new', id);
			}
			this.props.actions.sendSignal('offer_click_specific', id);
			this.props.actions.sendSignal('offer_dsp_session', id);
		}
	}

	/**
	 * Handle clicking on the close button
	 * @param  {Object} event the click event
	 */
	clickCloseButton(event) {
		// Prevent the event from propagating and linking to the Reward Detail
		event.preventDefault();
		event.stopPropagation();

		const { id } = this.props;
		this.props.actions.removeOffer(id);
		this.props.actions.sendSignal('remove_offer_link', id);
	}

	/**
	 * Helper render function for the expires text.
	 * @return {JSX} JSX for the Rewards Items List
	 */
	renderExpiresText() {
		const { expires } = this.props;
		const delta = computeTimeDelta(new Date(expires), new Date());
		return t('rewards_expires_in', [delta.count, t(`rewards_expires_in_${delta.type}`)]);
	}

	/**
	 * React's required render function. Returns JSX
	 * @return {JSX} JSX for rendering a Reward within the Rewards List
	 */
	render() {
		const {
			id,
			isLong,
			text,
			unread,
			disabled,
			logoUrl,
		} = this.props;
		const itemClassName = ClassNames('RewardListItem', 'row', {
			'RewardListItem--greyscale': disabled,
			'RewardListItem--unread': unread,
			'RewardListItem--elongated': isLong,
			'not-clickable': disabled,
			clickable: !disabled,
		});

		return (
			<Link to={`/detail/rewards/detail/${encodeURI(id)}`} className={itemClassName} onClick={this.handleClick}>
				<div className="flex-container align-middle full-height full-width">
					<div className="RewardListItem__image_container">
						<img className="RewardListItem__image" src={logoUrl} />
					</div>
					<div className="flex-grow">
						<div className="RewardListItem__text">{ text }</div>
						<div className="RewardListItem__expires">{ this.renderExpiresText() }</div>
					</div>
					<div className="flex-container align-justify flex-dir-column full-height">
						<div className="RewardListItem__close_button clickable" onClick={this.clickCloseButton}>
							<svg height="10" width="10" fillRule="evenodd">
								<path d="M6.719 5l2.923-2.924A1.216 1.216 0 0 0 7.924.356L5 3.281 2.076.356a1.216 1.216 0 1 0-1.72 1.72L3.28 5 .356 7.924a1.216 1.216 0 0 0 1.719 1.72L5 6.719l2.924 2.923a1.216 1.216 0 1 0 1.72-1.719L6.719 5z" />
							</svg>
						</div>
						<div className="RewardListItem__details_link">
							<svg height="16" width="10" fillRule="evenodd">
								<path d="M0 13.881l1.892 1.833L10 7.857 1.892 0 0 1.833l6.216 6.024z" />
							</svg>
						</div>
					</div>
				</div>
			</Link>
		);
	}
}

export default RewardListItem;
