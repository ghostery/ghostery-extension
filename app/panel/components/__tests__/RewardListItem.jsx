/**
 * Reward List Item Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';
import RewardListItem from '../BuildingBlocks/RewardListItem';

// Fake the translation function to only return the translation key
global.t = function (str) {
	return str;
};

describe('app/panel/components/BuildingBlocks/RewardListItem.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('reward list item is rendered correctly', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				disabled: false,
				expires: 1527885633645,
				id: 'test_reward_id',
				index: 0,
				isLong: false,
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				text: 'test reward title',
				unread: false
			};
			const component = renderer.create(
				<MemoryRouter>
					<RewardListItem {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward list item is rendered correctly when disabled', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				disabled: true,
				expires: 1527885633645,
				id: 'test_reward_id',
				index: 0,
				isLong: false,
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				text: 'test reward title',
				unread: false
			};
			const component = renderer.create(
				<MemoryRouter>
					<RewardListItem {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward list item is rendered correctly when unread', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				disabled: false,
				expires: 1527885633645,
				id: 'test_reward_id',
				index: 0,
				isLong: false,
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				text: 'test reward title',
				unread: true
			};
			const component = renderer.create(
				<MemoryRouter>
					<RewardListItem {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward list item is rendered correctly when elongated', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				disabled: false,
				expires: 1527885633645,
				id: 'test_reward_id',
				index: 0,
				isLong: true,
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				text: 'test reward title',
				unread: false
			};
			const component = renderer.create(
				<MemoryRouter>
					<RewardListItem {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('reward list item is rendered correctly when disabled, unread, and elongated', () => {
			const initialState = {
				actions: {
					updateRewardsData: () => {},
					toggleOffersEnabled: () => {},
					removeOffer: () => {},
					setOfferRead: () => {},
					sendSignal: () => {}
				},
				disabled: true,
				expires: 1527885633645,
				id: 'test_reward_id',
				index: 0,
				isLong: true,
				logoUrl: 'https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg',
				text: 'test reward title',
				unread: true
			};
			const component = renderer.create(
				<MemoryRouter>
					<RewardListItem {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});
});
