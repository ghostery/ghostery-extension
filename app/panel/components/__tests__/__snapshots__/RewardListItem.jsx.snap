// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel/components/BuildingBlocks/RewardListItem.jsx Snapshot tests with react-test-renderer reward list item is rendered correctly 1`] = `
<a
  className="RewardListItem row clickable"
  href="/detail/rewards/detail/test_reward_id"
  onClick={[Function]}
>
  <div
    className="flex-container align-middle full-height full-width"
  >
    <div
      className="RewardListItem__image_container"
    >
      <img
        className="RewardListItem__image"
        src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
      />
    </div>
    <div
      className="flex-grow"
    >
      <div
        className="RewardListItem__text"
      >
        test reward title
      </div>
      <div
        className="RewardListItem__expires"
      >
        rewards_expires_in
      </div>
    </div>
    <div
      className="flex-container align-justify flex-dir-column full-height"
    >
      <div
        className="RewardListItem__close_button clickable"
        onClick={[Function]}
      >
        <svg
          fillRule="evenodd"
          height="10"
          width="10"
        >
          <path
            d="M6.719 5l2.923-2.924A1.216 1.216 0 0 0 7.924.356L5 3.281 2.076.356a1.216 1.216 0 1 0-1.72 1.72L3.28 5 .356 7.924a1.216 1.216 0 0 0 1.719 1.72L5 6.719l2.924 2.923a1.216 1.216 0 1 0 1.72-1.719L6.719 5z"
          />
        </svg>
      </div>
      <div
        className="RewardListItem__details_link"
      >
        <svg
          fillRule="evenodd"
          height="16"
          width="10"
        >
          <path
            d="M0 13.881l1.892 1.833L10 7.857 1.892 0 0 1.833l6.216 6.024z"
          />
        </svg>
      </div>
    </div>
  </div>
</a>
`;

exports[`app/panel/components/BuildingBlocks/RewardListItem.jsx Snapshot tests with react-test-renderer reward list item is rendered correctly when disabled 1`] = `
<a
  className="RewardListItem row RewardListItem--greyscale not-clickable"
  href="/detail/rewards/detail/test_reward_id"
  onClick={[Function]}
>
  <div
    className="flex-container align-middle full-height full-width"
  >
    <div
      className="RewardListItem__image_container"
    >
      <img
        className="RewardListItem__image"
        src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
      />
    </div>
    <div
      className="flex-grow"
    >
      <div
        className="RewardListItem__text"
      >
        test reward title
      </div>
      <div
        className="RewardListItem__expires"
      >
        rewards_expires_in
      </div>
    </div>
    <div
      className="flex-container align-justify flex-dir-column full-height"
    >
      <div
        className="RewardListItem__close_button clickable"
        onClick={[Function]}
      >
        <svg
          fillRule="evenodd"
          height="10"
          width="10"
        >
          <path
            d="M6.719 5l2.923-2.924A1.216 1.216 0 0 0 7.924.356L5 3.281 2.076.356a1.216 1.216 0 1 0-1.72 1.72L3.28 5 .356 7.924a1.216 1.216 0 0 0 1.719 1.72L5 6.719l2.924 2.923a1.216 1.216 0 1 0 1.72-1.719L6.719 5z"
          />
        </svg>
      </div>
      <div
        className="RewardListItem__details_link"
      >
        <svg
          fillRule="evenodd"
          height="16"
          width="10"
        >
          <path
            d="M0 13.881l1.892 1.833L10 7.857 1.892 0 0 1.833l6.216 6.024z"
          />
        </svg>
      </div>
    </div>
  </div>
</a>
`;

exports[`app/panel/components/BuildingBlocks/RewardListItem.jsx Snapshot tests with react-test-renderer reward list item is rendered correctly when disabled, unread, and elongated 1`] = `
<a
  className="RewardListItem row RewardListItem--greyscale RewardListItem--unread RewardListItem--elongated not-clickable"
  href="/detail/rewards/detail/test_reward_id"
  onClick={[Function]}
>
  <div
    className="flex-container align-middle full-height full-width"
  >
    <div
      className="RewardListItem__image_container"
    >
      <img
        className="RewardListItem__image"
        src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
      />
    </div>
    <div
      className="flex-grow"
    >
      <div
        className="RewardListItem__text"
      >
        test reward title
      </div>
      <div
        className="RewardListItem__expires"
      >
        rewards_expires_in
      </div>
    </div>
    <div
      className="flex-container align-justify flex-dir-column full-height"
    >
      <div
        className="RewardListItem__close_button clickable"
        onClick={[Function]}
      >
        <svg
          fillRule="evenodd"
          height="10"
          width="10"
        >
          <path
            d="M6.719 5l2.923-2.924A1.216 1.216 0 0 0 7.924.356L5 3.281 2.076.356a1.216 1.216 0 1 0-1.72 1.72L3.28 5 .356 7.924a1.216 1.216 0 0 0 1.719 1.72L5 6.719l2.924 2.923a1.216 1.216 0 1 0 1.72-1.719L6.719 5z"
          />
        </svg>
      </div>
      <div
        className="RewardListItem__details_link"
      >
        <svg
          fillRule="evenodd"
          height="16"
          width="10"
        >
          <path
            d="M0 13.881l1.892 1.833L10 7.857 1.892 0 0 1.833l6.216 6.024z"
          />
        </svg>
      </div>
    </div>
  </div>
</a>
`;

exports[`app/panel/components/BuildingBlocks/RewardListItem.jsx Snapshot tests with react-test-renderer reward list item is rendered correctly when elongated 1`] = `
<a
  className="RewardListItem row RewardListItem--elongated clickable"
  href="/detail/rewards/detail/test_reward_id"
  onClick={[Function]}
>
  <div
    className="flex-container align-middle full-height full-width"
  >
    <div
      className="RewardListItem__image_container"
    >
      <img
        className="RewardListItem__image"
        src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
      />
    </div>
    <div
      className="flex-grow"
    >
      <div
        className="RewardListItem__text"
      >
        test reward title
      </div>
      <div
        className="RewardListItem__expires"
      >
        rewards_expires_in
      </div>
    </div>
    <div
      className="flex-container align-justify flex-dir-column full-height"
    >
      <div
        className="RewardListItem__close_button clickable"
        onClick={[Function]}
      >
        <svg
          fillRule="evenodd"
          height="10"
          width="10"
        >
          <path
            d="M6.719 5l2.923-2.924A1.216 1.216 0 0 0 7.924.356L5 3.281 2.076.356a1.216 1.216 0 1 0-1.72 1.72L3.28 5 .356 7.924a1.216 1.216 0 0 0 1.719 1.72L5 6.719l2.924 2.923a1.216 1.216 0 1 0 1.72-1.719L6.719 5z"
          />
        </svg>
      </div>
      <div
        className="RewardListItem__details_link"
      >
        <svg
          fillRule="evenodd"
          height="16"
          width="10"
        >
          <path
            d="M0 13.881l1.892 1.833L10 7.857 1.892 0 0 1.833l6.216 6.024z"
          />
        </svg>
      </div>
    </div>
  </div>
</a>
`;

exports[`app/panel/components/BuildingBlocks/RewardListItem.jsx Snapshot tests with react-test-renderer reward list item is rendered correctly when unread 1`] = `
<a
  className="RewardListItem row RewardListItem--unread clickable"
  href="/detail/rewards/detail/test_reward_id"
  onClick={[Function]}
>
  <div
    className="flex-container align-middle full-height full-width"
  >
    <div
      className="RewardListItem__image_container"
    >
      <img
        className="RewardListItem__image"
        src="https://www.ghostery.com/wp-content/themes/ghostery/images/ghostery_logo.svg"
      />
    </div>
    <div
      className="flex-grow"
    >
      <div
        className="RewardListItem__text"
      >
        test reward title
      </div>
      <div
        className="RewardListItem__expires"
      >
        rewards_expires_in
      </div>
    </div>
    <div
      className="flex-container align-justify flex-dir-column full-height"
    >
      <div
        className="RewardListItem__close_button clickable"
        onClick={[Function]}
      >
        <svg
          fillRule="evenodd"
          height="10"
          width="10"
        >
          <path
            d="M6.719 5l2.923-2.924A1.216 1.216 0 0 0 7.924.356L5 3.281 2.076.356a1.216 1.216 0 1 0-1.72 1.72L3.28 5 .356 7.924a1.216 1.216 0 0 0 1.719 1.72L5 6.719l2.924 2.923a1.216 1.216 0 1 0 1.72-1.719L6.719 5z"
          />
        </svg>
      </div>
      <div
        className="RewardListItem__details_link"
      >
        <svg
          fillRule="evenodd"
          height="16"
          width="10"
        >
          <path
            d="M0 13.881l1.892 1.833L10 7.857 1.892 0 0 1.833l6.216 6.024z"
          />
        </svg>
      </div>
    </div>
  </div>
</a>
`;
